core/tests/unit/api_core/test_retry.py

```diff
@@ -82,6 +82,34 @@ def target():
     'google.api.core.helpers.datetime_helpers.utcnow',
     return_value=datetime.datetime.min,
     autospec=True)
+def test_retry_target_w_on_error(utcnow, sleep):
+    predicate = retry.if_exception_type(ValueError)
+    call_count = {'target': 0, 'on_error': 0}
+
+    def on_error():
+        call_count['on_error'] += 1
+
+    def target():
+        call_count['target'] += 1
+        if call_count['target'] < 3:
+            raise ValueError()
+        return 42
+
+    result = retry.retry_target(
+        target, predicate, range(10), None, on_error=on_error)
+
+    assert result == 42
+    assert call_count['target'] == 3
+    assert call_count['on_error'] == 2
+
+    sleep.assert_has_calls([mock.call(0), mock.call(1)])
```

Comment:
Also check the `call_count == 2`?

Reply:
That assertion has so little obvious value that I just ignored it.